src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/ControlProblemElements.h

```diff
@@ -0,0 +1,745 @@
+ * element that will be embedded as cost function or as equality constraint.
+ * The element is described by \f$ A \f$ and \f$ b \f$. \f$ A \f$ is the
+ * element matrix and \f$ x \f$ and \f$ b \f$ the element vector.
+ * In case of cost function \f$ A \f$ and \f$ b \f$ represents:
+ * \f$ \norm{Ax - b}^2 \f$. In case of equality constraint
+ * \f$ Ax = b \f$
+ */
+class CostFunctionOrEqualityConstraintElement : public ControlProblemElement
+{
+protected:
+    iDynTree::VectorDynSize m_b; /**< Element Vector */
+
+public:
+    /**
+     * Constructor
+     * @param kinDyn KinDynComputations object
+     */
+    CostFunctionOrEqualityConstraintElement(iDynTree::KinDynComputations& kinDyn);
+
+    /**
+     * Get the element vector
+     * @return the element vector
+     */
+    virtual const iDynTree::VectorDynSize& getB() = 0;
+};
+
+/**
+ * InequalityConstraintElement describes a control problem
+ * element that will be embedded as inequality constraint.
+ * The element is described by \f$ A \f$, \f$ l \f$ and  \f$ u \f$.
+ * \f$ l \le Ax \le u \f$
+ */
+class InequalityConstraintElement : public ControlProblemElement
+{
+protected:
+    iDynTree::VectorDynSize m_l; /**< Lower bound */
+    iDynTree::VectorDynSize m_u; /**< Upper bound */
+
+public:
+    /**
+     * Constructor
+     * @param kinDyn KinDynComputations object
+     */
+    InequalityConstraintElement(iDynTree::KinDynComputations& kinDyn);
+
+    /**
+     * Get the upper bound
+     * @return the upper bound
+     */
+    virtual const iDynTree::VectorDynSize& getUpperBound() = 0;
+
+    /**
+     * Get the lower bound
+     * @return the lower bound
+     */
+    virtual const iDynTree::VectorDynSize& getLowerBound() = 0;
+};
+
+/**
+ * CartesianElement describes a Cartesian element that will be embedded as
+ * equality constraint or cost function.
+ * The Cartesian element in useful to control the position/orientation/pose of a frame/CoM of the
+ * robot
+ */
+class CartesianElement : public CostFunctionOrEqualityConstraintElement
+{
+public:
+    /** Cartesian Element Type */
+    enum class Type
+    {
+        POSE,
+        POSITION,
+        ORIENTATION,
+        ONE_DIMENSION
+    };
+
+    /** Name of the axis controlled in case of ONE_DIMENSION element is chosen */
+    enum class AxisName
+    {
+        X,
+        Y,
+        Z
+    };
+
+private:
+    /** Base acceleration index */
+    iDynTree::IndexRange m_baseAccelerationIndex{iDynTree::IndexRange::InvalidRange()};
+    /** Joint acceleration index */
+    iDynTree::IndexRange m_jointAccelerationIndex{iDynTree::IndexRange::InvalidRange()};
+    /** Control type index */
+    iDynTree::IndexRange m_typeIndex{iDynTree::IndexRange::InvalidRange()};
+
+    std::unique_ptr<PositionPID> m_positionPID; /**< 3D Position PID. */
+    std::unique_ptr<OrientationPID> m_orientationPID; /**< Orientation PID. */
+    std::unique_ptr<OneDegreePID> m_oneDegreePID; /**< One Degree of freedom PID. */
```

Comment:
Are you planning to change the PID type at runtime? Otherwise, why not defining three different classes? Or abstract the concept of PID?